src/liballoc/raw_vec.rs

```diff
@@ -350,19 +351,25 @@ impl<T> RawVec<T> {
 
             // NOTE: we don't early branch on ZSTs here because we want this
             // to actually catch "asking for more than usize::MAX" in that case.
-            // If we make it past the first branch then we are guaranteed to
-            // panic.
+            let required_cap = used_cap.checked_add(needed_extra_cap)
+               .expect("capacity overflow");
+
+            // Shortcut for ZST to simplify further computations.
+            if elem_size == 0 {
+                return;
+            }
 
             // Don't actually need any more capacity.
-            // Wrapping in case they give a bad `used_cap`
-            if self.cap().wrapping_sub(used_cap) >= needed_extra_cap {
+            if required_cap <= self.cap {
                 return;
             }
 
-            // Nothing we can really do about these checks :(
-            let new_cap = used_cap.checked_add(needed_extra_cap)
-                                  .and_then(|cap| cap.checked_mul(2))
-                                  .expect("capacity overflow");
+            // Cannot overflow, because cap <= isize::MAX, and type of cap is usize.
+            let double_cap = self.cap * 2;
```

Comment:
This is fine, same as before, but capacity really must not exceed `isize::MAX`. I guess we used to & do rely on the allocator to OOM to save us.

I'm not sure that's sound for the element type u8 on a 32-bit platform, could we end up with a successful allocation there, larger than isize::MAX?